python/ray/serve/backend_state.py

```diff
@@ -248,19 +338,16 @@ def __init__(self, controller_name: str, detached: bool,
             BackendReplica]]] = defaultdict(lambda: defaultdict(list))
         self._backend_metadata: Dict[BackendTag, BackendInfo] = dict()
         self._target_replicas: Dict[BackendTag, int] = defaultdict(int)
+        self._backend_goals: Dict[BackendTag, GoalId] = dict()
         self._target_versions: Dict[BackendTag, str] = dict()
-        self.backend_goals: Dict[BackendTag, GoalId] = dict()
-
-        # Un-checkpointed state.
-        self.pending_goals: Dict[GoalId, asyncio.Event] = dict()
 
         checkpoint = self._kv_store.get(CHECKPOINT_KEY)
         if checkpoint is not None:
             (self._replicas, self._backend_metadata, self._target_replicas,
-             self._target_versions, self.backend_goals,
-             pending_goal_ids) = pickle.loads(checkpoint)
+             self._target_versions,
+             self._backend_goals) = pickle.loads(checkpoint)
```

Comment:
I'm struggling to remember the exact reason why we used to keep these separately.


The only difference I see is that the Goal Manager will have both the old Goal and the new Goal, whereas if we checkpoint `self._backend_goals`, we will only have the new Goal.


I guess the case of concern is the following:
```
new_goal_id, existing_goal_id = self._set_backend_goal(...)
self._checkpoint()    # The existing_goal_id would not exist in the checkpoint anymore
# Crash occurs here
if existing_goal_id is not None:
    self._goal_manager.complete_goal(existing_goal_id)
return new_goal_id
```
But we are still fine here because the goal manager will just return if the `goal_id` passed into [`wait_for_goal(goal_id)`](https://github.com/ray-project/ray/blob/34b5781ae0d46b769e8bac400e0cbad6c60c8fcd/python/ray/serve/async_goal_manager.py#L34) doesn't exist

Reply:
TL;DR I think this is fine.